server/src/test/kotlin/org/itmodreamteam/myrest/server/service/restaurant/ReservationTest.kt

```diff
@@ -0,0 +1,188 @@
+        authorize(user2)
+        val reservation = reservationService.submitReservationForApproval(
+            table,
+            reservation.activeUntil,
+            reservation.activeUntil.plusHours(2)
+        )
+        Assertions.assertThat(reservation).isNotNull
+        Assertions.assertThat(reservation.status).isEqualTo(ReservationStatus.PENDING)
+        Assertions.assertThat(reservation.user).isEqualTo(user2)
+        Assertions.assertThat(reservation.table).isEqualTo(table)
+    }
+
+    @Test(expected = UserException::class)
+    fun `Given existing reservation, when reserve the same table for overlapping interval, then failure`() {
+        authorize(user2)
+        reservationService.submitReservationForApproval(
+            table,
+            reservation.activeFrom.minusHours(2),
+            reservation.activeFrom.plusMinutes(1)
+        )
+    }
+
+    @Test
+    fun `Given rejected existing reservation, when reserve the same table for overlapping interval, then a reservation created`() {
+        authorize(user1)
+        reservationService.reject(reservation)
+        reservationService.submitReservationForApproval(table, reservation.activeFrom, reservation.activeUntil)
+    }
+
+    @Test(expected = UserException::class)
+    fun `When try to reject other user's reservation, then failure`() {
+        authorize(user2)
+        reservationService.reject(reservation)
+    }
+
+    @Test
+    fun `Given existing reservation in state PENDING, when manager approves it, then reservation is APPROVED`() {
+        authorize(userReservationManager)
+        val approved = reservationService.approve(reservation)
+        Assertions.assertThat(approved.status).isEqualTo(ReservationStatus.APPROVED)
+    }
+
+    @Test(expected = UserException::class)
+    fun `Given existing PENDING reservation and manager who is not manager of the related restaurant, when try to approve it, then failure`() {
+        authorize(userManagerOfAnotherRestaurant)
+        reservationService.approve(reservation)
+    }
+
+    @Test(expected = UserException::class)
+    fun `Given existing PENDING reservation and user who is not manager of any restaurant, when try to approve it, then failure`() {
+        authorize(user1)
+        reservationService.approve(reservation)
+    }
+
+    // TODO overwrite with nested tests after migration to junit5
+    @Test(expected = UserException::class)
+    fun `Given existing APPROVED reservation and user who is not manager of any restaurant, when try to start it, then failure`() {
+        authorize(userReservationManager)
+        val approved = reservationService.approve(reservation)
+
+        authorize(user1)
+        reservationService.start(approved)
+    }
+
+    @Test(expected = UserException::class)
+    fun `Given existing APPROVED reservation and user who is manager of related restaurant but is not responsible for reservation, when try to start it, then failure`() {
```

Comment:
I believe **any** of the restaurant managers should be able to start reservation because the reservation could be approved a couple of days ago, and currently that manager could be having a day off or something...